### DISCUSSION TITLE
Two ways to compile?

### DISCUSSION THREAD

**Discussion by schnapsy** [nohal/nsk_pi#15] [Q&A]
Sorry to bother you...

I'm trying to compile - unsuccessfully - the plugin with the DBT phrase.
Error messages every time.

So, to check, I tried to compile the plugin without modifying it (no addition of additional nmea0183 sentences).

1. Compilation, with the specific NSK method described here (https://github.com/nohal/nsk_pi#readme) does not work.
I actually get the file under:
C:\Users\alain\AppData\Local\OCNplugins\AS_nsk_pi_AS\build\app\NSK-0.2-msvc-wx32-10\plugins\NSK_pi\
C:\Users\alain\AppData\Local\OCNplugins\AS_nsk_pi_AS\build\app\NSK-0.2-msvc-wx32-10\plugins\NSK_pi.dll
By copying/pasting the folder and the dll, OCN starts and stops immediately.

2. BUT, compiling with the "classic" method described here works perfectly (Third-party plugin build example - https://gist.github.com/nohal/63d9eef725477a8f9611ddf5182c0f50).
I get :
C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\app\NSK-0.2-msvc-wx32-10.0.19045\plugins\NSK_pi\
C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\app\NSK-0.2-msvc-wx32-10.0.19045\plugins\NSK_pi.dll
By copying/pasting the folder and the dll, OCN starts and the plugin works, even after restarting the PC.

The visible difference lies in the title of the file:
NSK-0.2-msvc-wx32-10 vs. NSK-0.2-msvc-wx32-10.0.19045
See if this is the case for others...
____________________________________________________________________________

Incidentally, if you have the time, can you tell me where the error lies in the nsk.cpp program to get the depth display, knowing that all the other addons (declarations and nsk.h) have been written .

//--------- ADD DBT AS ------------------

void NSK::ProcessSentence(std::unique_ptr<marnav::nmea::dbt> s,
     rapidjson::Value& values_array,
     rapidjson::Document::AllocatorType& allocator)
{
     Valuedbt(kObjectType);
     dbt.AddMember("path", "environment.depth.belowTransducer", allocator);
     dbt.AddMember("value", s->get_depth_meter(), allocator);
     //val.AddMember("value", s->get_depth_meter(2), allocator);
     values_array. PushBack(dbt, allocator);
}
//-----------------------------------------

**Comment by nohal** [comment]
What exact error do you see with your code? I see an error in the very first line, but that likely is you not using formatting and part of the text being lost.

**Reply by schnapsy** [reply]
Thank you for your quick reply.
I changed the formatting, but it didn't change anything.

Here are the error lines that appear in red at the end of the compilation:   
_________________________________________________________________________________________________
Génération de code en cours...
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: La commande "setlocal [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: cmake --build C:/Users/alain/AppData/Local/OCNplugins/nsk_pi/build --parallel 4 --config Release --target
 NSK_pi [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: if %errorlevel% neq 0 goto :cmEnd [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build.vcx
proj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: :cmEnd [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi
\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: :cmErrorLevel [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: exit /b %1 [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: :cmDone [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: if %errorlevel% neq 0 goto :VCEnd [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build.vcx
proj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: :VCEnd" s'est arrêtée avec le code 1. [C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build\tarball-build
.vcxproj]

C:\Users\alain\AppData\Local\OCNplugins\nsk_pi\build>

**Reply by nohal** [reply]
There is no error in this snippet. BTW, I have implemented the `DBT` sentence in https://github.com/nohal/nsk_pi/commit/07bbd16c7930c98fd969528dbca1e61fb6b9153a so you may compare to your code to see what your problem is.

**Reply by schnapsy** [reply]
Thanks Nohal.
I found and understood where my mistake was.
I wrote :
_dbt.AddMember("value", s->get_depth_meter(), allocator);_
instead of :
_dbt.AddMember("value", s->get_depth_meter()**.value().value()**, allocator);_

**Reply by nohal** [reply]
`marnav` uses some modern C++ features like `std::optional`, you should have a look how they work (and of course at the code of the already implemented NMEA sentences)
